### PR TITLE
Initial set of changes for Change Log Categorization.

### PR DESCRIPTION
**Pull request by TheLayman**

### CODE REVIEW

File: Tasks/GitHubReleaseV0/main.ts [hunk 1]
```diff
@@ -101,13 +101,22 @@ class Main {
         const releaseNotesFile = tl.getPathInput(Inputs.releaseNotesFile, false, true);
         const releaseNoteInput = tl.getInput(Inputs.releaseNotes);
         const showChangeLog: boolean = tl.getBoolInput(Inputs.addChangeLog);
-
+        const changeLogLabelsInput = tl.getInput(Inputs.changeLogLabels);
+        let changeLogLabels: string;
+        try{
+            changeLogLabels = JSON.parse(changeLogLabelsInput);
```
Comment (shyama1208) [inline Tasks/GitHubReleaseV0/main.ts]: whats the value for changeLogLabelsInput when user does not select any value?

File: Tasks/GitHubReleaseV0/main.ts [hunk 2]
```diff
@@ -101,13 +101,22 @@ class Main {
...
+        }
+        catch(error){
+            changeLogLabels = null;
+            console.log(tl.loc("ChangeLogLabelsParseError"));
```
Comment (shyama1208) [inline Tasks/GitHubReleaseV0/main.ts]: make sure this is not logged when user has not selected any category

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 3]
```diff
@@ -15,7 +15,7 @@ export class ChangeLog {
      * @param compareWithRelease
      * @param changeLogCompareToReleaseTag
      */
-    public async getChangeLog(githubEndpointToken: string, repositoryName: string, target: string, top: number, compareWithRelease: ChangeLogStartCommit, changeLogCompareToReleaseTag?: string): Promise<string> {
+    public async getChangeLog(githubEndpointToken: string, repositoryName: string, target: string, top: number, compareWithRelease: ChangeLogStartCommit, changeLogCompareToReleaseTag?: string, changeLogLabels?: any): Promise<string> {
```
Comment (shyama1208) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: why type is any?

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 4]
```diff
@@ -315,6 +365,70 @@ export class ChangeLog {
...
+        return Delimiters.newLine + Delimiters.newLine + Delimiters.hash + Delimiters.hash + Delimiters.space + labelDetails.displayName + Delimiters.newLine + Delimiters.newLine;
+    }
+
+    /**
+     * Returns the log for a single issue.
+     * Log format: * #issueId: issueTitle
+     * @param issueDetails 
+     */
+    private _getChangeLogPerIssue(issueDetails){
+        return Delimiters.star + Delimiters.space + Delimiters.hash + issueDetails.issueId + Delimiters.colon + Delimiters.space + issueDetails.issue;
```
Comment (shyama1208) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: make this a parameterized string and then use.

File: Tasks/GitHubReleaseV0/task.json [hunk 5]
```diff
@@ -219,6 +219,14 @@
...
             "visibleRule": "changeLogCompareToRelease = lastNonDraftReleaseByTag",
             "helpMarkDown": "Specify the regex for release tag. Release matching this tag will be used as base for changelog computation."
+        },
+        {
+            "name": "changeLogLabels",
+            "type": "multiLine",
+            "label": "Labels",
+            "required": false,
+            "groupName": "changeLogConfiguration",
+            "helpMarkDown": "Specify the labels to be used for changelog categorization. If leaved blank, ChangeLog is generated based on commit history."
```
Comment (shyama1208) [inline Tasks/GitHubReleaseV0/task.json]: If left blank..... Is this string reviewed by PM?

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts (hunk 3 above)
Comment (TheLayman) [reply Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: User input is a string which is then JSON parsed.

File: Tasks/GitHubReleaseV0/task.json [hunk 6]
```diff
@@ -219,6 +219,30 @@
...
+                "commitBased": "Commit based",
+                "issueBased": "Issue based"
+            }
+        },
+        {
+            "name": "changeLogLabels",
+            "type": "multiLine",
+            "label": "Labels",
+            "required": true,
+            "visibleRule": "changeLogType = issueBased",
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/task.json]: there should be a default value. We should not have the labels key as it is redundant

File: Tasks/GitHubReleaseV0/task.json [hunk 7]
```diff
@@ -219,6 +219,30 @@
             "groupName": "changeLogConfiguration",
             "visibleRule": "changeLogCompareToRelease = lastNonDraftReleaseByTag",
             "helpMarkDown": "Specify the regex for release tag. Release matching this tag will be used as base for changelog computation."
+        },
+        
+        {
+            "name": "changeLogType",
+            "type": "radio",
+            "label": "Change log type",
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/task.json]: In display names, is Changelog one word or split. Need to be consistent

File: Tasks/GitHubReleaseV0/task.loc.json [hunk 8]
```diff
@@ -219,6 +219,29 @@
...
+      "name": "changeLogType",
+      "type": "radio",
+      "label": "ms-resource:loc.input.label.changeLogType",
+      "required": true,
+      "defaultValue": "issueBased",
+      "groupName": "changeLogConfiguration",
+      "visibleRule": "addChangeLog = true",
+      "helpMarkDown": "ms-resource:loc.input.help.changeLogType",
+      "options": {
+        "commitBased": "Commit based",
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/task.loc.json]: how do options get localized?

File: Tasks/GitHubReleaseV0/Strings/resources.resjson/en-US/resources.resjson [hunk 9]
```diff
@@ -112,5 +116,9 @@
   "loc.messages.TagRequiredCreateAction": "Tag source is set to manual- please specify a tag for create action. For yaml syntax see: https://aka.ms/AA3m1bq",
   "loc.messages.NoFileFoundMatchingPattern": "No files found matching '%s'. Nothing to upload.",
   "loc.messages.PatternIsADirectory": "'%s' cannot be uploaded as it is a directory. Please specify a file.",
-  "loc.messages.SearchingFileMatchingPattern": "Searching for file(s) matching '%s'."
+  "loc.messages.SearchingFileMatchingPattern": "Searching for file(s) matching '%s'.",
+  "loc.messages.IssuesFetchError": "Error fetching issues. Cannot generate change log.",
+  "loc.messages.NoIssuesLinkedError": "No issues are linked to commits in the specified commit Diff.",
+  "loc.messages.ChangeLogLabelsParseError": "Unable to parse changeLog labels.",
+  "loc.messages.NoLabelsSpecified": "Error occured while parsing the labels. For yaml syntax see: https://aka.ms/AA3m1bq"
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/Strings/resources.resjson/en-US/resources.resjson]: LabelSyntaxError

File: Tasks/GitHubReleaseV0/main.ts [hunk 10]
```diff
@@ -97,18 +97,34 @@ class Main {
...
         Utility.validateReleaseNotesSource(releaseNotesSource);
-
         const releaseNotesFile = tl.getPathInput(Inputs.releaseNotesFile, false, true);
         const releaseNoteInput = tl.getInput(Inputs.releaseNotes);
         const showChangeLog: boolean = tl.getBoolInput(Inputs.addChangeLog);
+        let changeLog: string = "";
+        if (showChangeLog){
+            let changeLogLabels: any;
+            const changeLogCompareToRelease = tl.getInput(Inputs.changeLogCompareToRelease);
+            const changeLogLabelsInput = tl.getInput(Inputs.changeLogLabels);
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/main.ts]: nit: this could move into the issueBased if block

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 11]
```diff
@@ -60,30 +60,157 @@ export class ChangeLog {
...
+    }
+     /**
+     * Generate issue based ChangeLog
+     * @param issues
+     * @param repositoryName 
+     * @param release 
+     * @param githubEndpointToken
+     * @param labels
+     */
+    private async _getIssueBasedChangeLog(issues: number[], repositoryName: string, release: Release, githubEndpointToken: string, labels: any[]) {
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: nit: potential refactoring for see more logic across the three methods

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 12]
```diff
@@ -458,9 +699,12 @@ export class ChangeLog {
     // https://github.com/moby/moby/commit/df23a1e675c7e3cbad617374d85c48103541ee14?short_path=6206c94#diff-6206c94cde21ec0a5563c8369b71e609
     // Supported format for GitHub issues: #26 GH-26 repositoryName#26 repositoryNameGH-26, where GH is case in-sensitive.
     private readonly _issueRegex = new RegExp("(?:^|[^A-Za-z0-9_]?)([a-z0-9_]+/[a-zA-Z0-9-_.]+)?(?:#|[G|g][H|h]-)([0-9]+)(?:[^A-Za-z_]|$)", "gm");
-    private readonly _ChangeLogTitle: string = "\n\n## Changes:\n\n";
+    private readonly _changeLogTitle: string = "Changes";
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: this should likely come from resrouces

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 13]
```diff
@@ -458,9 +699,12 @@ export class ChangeLog {
...
     private readonly _seeMoreText: string = "See more";
+    private readonly _noStateSpecified: string = "none";
+    private readonly _defaultGroup: string = "Others";
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: display names like See more, Others should come from resources

**Pull request merged**